### PR TITLE
KNOX-1162 - Logging stacktrace for FATAL messages and displaying a meaningful error message in case of missing/non-parsable JAAS configuration

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?

When there is a misconfiguration in the supplied JAAS configuration (i.e. `conf/krb5JAASLogin.conf`) the server fails to start and the information in gateway.log wasn't at all helpful. To make it better the following changes have been made:
- *all* `FATAL` messages are logged with the full stack trace
- in case the supplied JAAS configuration file does not exist or cannot be parsed we wrap the `IOException` coming from security login configuration into our own `ConfigurationException` with a meaningful error message

(an additional change is to fix a unit test case in `MatcherTest` to avoid compilation error; I'm not sure how it was working before but it constantly fails locally for me)

## How was this patch tested?

Added new unit test cases and executed them (including integration tests):
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 18:18 min (Wall Clock)
[INFO] Finished at: 2019-02-15T23:48:43+01:00
[INFO] Final Memory: 265M/1643M
[INFO] ------------------------------------------------------------------------
```

In addition to unit testing the following manual test has been executed:

1. stopped the gateway server
2. built and deployed the new version of the gateway server (with my changes)
3. updated `/etc/knox/conf/krb5JAASLogin.conf`: removed surrounding quotes from the `keytab` property (which is invalid)
4. tried to start the gateway. As expected it failed to start
5. checked the log file if the enhanced error message appeared:

```
2019-02-15 23:05:44,047 FATAL knox.gateway (GatewayServer.java:main(168)) - Failed to start gateway: org.apache.knox.gateway.services.ServiceLifecycleException: Error while configuring registry authentication
org.apache.knox.gateway.services.ServiceLifecycleException: Error while configuring registry authentication
        at org.apache.knox.gateway.service.config.remote.zk.CuratorClientService.init(CuratorClientService.java:81)
        at org.apache.knox.gateway.services.DefaultGatewayServices.init(DefaultGatewayServices.java:79)
        at org.apache.knox.gateway.GatewayServer.main(GatewayServer.java:159)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.knox.gateway.launcher.Invoker.invokeMainMethod(Invoker.java:68)
        at org.apache.knox.gateway.launcher.Invoker.invoke(Invoker.java:39)
        at org.apache.knox.gateway.launcher.Command.run(Command.java:99)
        at org.apache.knox.gateway.launcher.Launcher.run(Launcher.java:75)
        at org.apache.knox.gateway.launcher.Launcher.main(Launcher.java:52)
Caused by: org.apache.knox.gateway.config.ConfigurationException: Error while getting secure configuration. This error usually indicates an issue within the supplied JAAS configuration: /etc/knox/conf/krb5JAASLogin.conf
        at org.apache.knox.gateway.service.config.remote.zk.RemoteConfigurationRegistryJAASConfig.<init>(RemoteConfigurationRegistryJAASConfig.java:71)
        at org.apache.knox.gateway.service.config.remote.zk.RemoteConfigurationRegistryJAASConfig.configure(RemoteConfigurationRegistryJAASConfig.java:61)
        at org.apache.knox.gateway.service.config.remote.zk.CuratorClientService.init(CuratorClientService.java:79)
        ... 11 more
Caused by: java.lang.SecurityException: java.io.IOException: Configuration Error:
        Line 7: expected [option key]
        at sun.security.provider.ConfigFile$Spi.<init>(ConfigFile.java:137)
        at sun.security.provider.ConfigFile.<init>(ConfigFile.java:102)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at java.lang.Class.newInstance(Class.java:442)
        at javax.security.auth.login.Configuration$2.run(Configuration.java:255)
        at javax.security.auth.login.Configuration$2.run(Configuration.java:247)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.login.Configuration.getConfiguration(Configuration.java:246)
        at org.apache.knox.gateway.service.config.remote.zk.RemoteConfigurationRegistryJAASConfig.<init>(RemoteConfigurationRegistryJAASConfig.java:66)
        ... 13 more
Caused by: java.io.IOException: Configuration Error:
        Line 7: expected [option key]
        at sun.security.provider.ConfigFile$Spi.ioException(ConfigFile.java:666)
        at sun.security.provider.ConfigFile$Spi.match(ConfigFile.java:572)
        at sun.security.provider.ConfigFile$Spi.parseLoginEntry(ConfigFile.java:477)
        at sun.security.provider.ConfigFile$Spi.readConfig(ConfigFile.java:427)
        at sun.security.provider.ConfigFile$Spi.init(ConfigFile.java:329)
        at sun.security.provider.ConfigFile$Spi.init(ConfigFile.java:271)
        at sun.security.provider.ConfigFile$Spi.<init>(ConfigFile.java:135)
        ... 24 more
```

### CODE REVIEW

**Comment by smolnar82** [comment]
@risdenk @rlevas @lmccay could you please review my changes? Thanks!

**Comment by smolnar82** [comment]
Restart build

File: gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java [hunk 1]
```diff
@@ -49,17 +51,26 @@
...
 
     static RemoteConfigurationRegistryJAASConfig configure(List<RemoteConfigurationRegistryConfig> configs, AliasService aliasService) {
         return new RemoteConfigurationRegistryJAASConfig(configs, aliasService);
     }
 
     private RemoteConfigurationRegistryJAASConfig(List<RemoteConfigurationRegistryConfig> configs, AliasService aliasService) {
+        try {
+          delegate = Configuration.getConfiguration();
+        } catch(Exception e) {
+          //populate the original error with a meaningful message; logging will happen later in the upper in the call hierarchy
```
Comment (rlevas) [inline gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java]: Change to:
"//populate the original error with a meaningful message; logging will happen later in the call hierarchy"

Comment (smolnar82) [reply gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java]: Fixed; thanks for the review!

File: gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java [hunk 2]
```diff
@@ -49,17 +51,26 @@
...
         return new RemoteConfigurationRegistryJAASConfig(configs, aliasService);
     }
 
     private RemoteConfigurationRegistryJAASConfig(List<RemoteConfigurationRegistryConfig> configs, AliasService aliasService) {
+        try {
+          delegate = Configuration.getConfiguration();
+        } catch(Exception e) {
+          //populate the original error with a meaningful message; logging will happen later in the call hierarchy
+          final String message = String.format(Locale.ROOT, "Error while getting secure configuration. This error usually indicates an issue within the supplied JAAS configuration: %s",
+              System.getProperty(GatewayConfig.KRB5_LOGIN_CONFIG, "Undefined"));
```
Comment (risdenk) [inline gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java]: Is this always a system property?

File: gateway-server/src/main/java/org/apache/knox/gateway/GatewayMessages.java [hunk 3]
```diff
@@ -34,13 +34,13 @@
 public interface GatewayMessages {
 
   @Message( level = MessageLevel.FATAL, text = "Failed to parse command line: {0}" )
-  void failedToParseCommandLine( @StackTrace( level = MessageLevel.DEBUG ) ParseException e );
+  void failedToParseCommandLine( @StackTrace( level = MessageLevel.FATAL ) ParseException e );
```
Comment (risdenk) [inline gateway-server/src/main/java/org/apache/knox/gateway/GatewayMessages.java]: Is this change necessary? I would have assumed that the MessageLevel was a minimum level to expose not the exact level.

File: gateway-service-remoteconfig/pom.xml [hunk 4]
```diff
@@ -38,6 +38,10 @@
             <groupId>org.apache.knox</groupId>
             <artifactId>gateway-spi</artifactId>
         </dependency>
+        <dependency>
```
Comment (risdenk) [inline gateway-service-remoteconfig/pom.xml]: Is this just for `ConfigurationException`?

File: gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfigTest.java [hunk 5]
```diff
@@ -36,6 +47,14 @@
 
 public class RemoteConfigurationRegistryJAASConfigTest {
 
+    @Rule
+    public final TemporaryFolder testFolder = new TemporaryFolder();
+
+    @Rule
+    public final ExpectedException expectedException = ExpectedException.none();
+
+    private static final String JAAS_CONFIG_ERRROR_PREFIX = "Error while getting secure configuration. This error usually indicates an issue within the supplied JAAS configuration";
```
Comment (risdenk) [inline gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfigTest.java]: This should be in the class `RemoteConfigurationRegistryJAASConfig` so it can be used in the test and in the actual throwing of the exception.

File: gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java [hunk 6]
```diff
@@ -796,7 +796,7 @@ public void testMultipleDoubleStarPathMatching() throws URISyntaxException {
     Template template;
     Template input;
     Matcher<String> stringMatcher;
-    Matcher<?>.Match match;
```
Comment (risdenk) [inline gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java]: Is this change necessary?

File: gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java (hunk 2 above)
Comment (smolnar82) [reply gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java]: Yepp; this is set before we try to start the gateway in `org.apache.knox.gateway.GatewayServer.configureKerberosSecurity(GatewayConfig)`

File: gateway-server/src/main/java/org/apache/knox/gateway/GatewayMessages.java (hunk 3 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/GatewayMessages.java]: The JIRA requested that there should be no FATAL messages displayed on DEBUG level (which I agree with); so I believe this is necessary. Any objection?

File: gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfigTest.java (hunk 5 above)
Comment (smolnar82) [reply gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfigTest.java]: Agree; will change it soon...

File: gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java (hunk 6 above)
Comment (smolnar82) [reply gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java]: Like I indicated in the description without this change I was not able to run tests locally. I'm not sure how it was working before in Travis CI..

Comment (risdenk) [reply gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java]: Yea do you by chance have the Java level not set to 7+ or something where it doesn't like the generics?

File: gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java (hunk 2 above)
Comment (risdenk) [reply gateway-service-remoteconfig/src/main/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfig.java]: Ah thanks

File: gateway-server/src/main/java/org/apache/knox/gateway/GatewayMessages.java (hunk 3 above)
Comment (risdenk) [reply gateway-server/src/main/java/org/apache/knox/gateway/GatewayMessages.java]: Ok makes sense. Didn't tie the two pieces together. The stacktrace level is basically opposite of what I expected.

File: gateway-service-remoteconfig/pom.xml (hunk 4 above)
Comment (smolnar82) [reply gateway-service-remoteconfig/pom.xml]: Yes. If there is an issue with the supplied JAAS config it is a `ConfigurationException`, right?
Moreover `ConfigurationException` is not a checked one so that I did not have to modify the API. I think it would be an exaggeration to create a new type of exception just for this purpose if we already have to proper one. Agreed?

Comment (risdenk) [reply gateway-service-remoteconfig/pom.xml]: Yea just checking that we didn't pull in a new dependency for another reason.

File: gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java (hunk 6 above)
Comment (smolnar82) [reply gateway-util-urltemplate/src/test/java/org/apache/knox/gateway/util/urltemplate/MatcherTest.java]:
```
$ java -version
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
```
I use the same version in my IDE too...

The same issue like described here: https://stackoverflow.com/questions/29621138/why-is-assertthatmap1-sameinstancemap2-not-allowed

File: gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfigTest.java (hunk 5 above)
Comment (smolnar82) [reply gateway-service-remoteconfig/src/test/java/org/apache/knox/gateway/service/config/remote/zk/RemoteConfigurationRegistryJAASConfigTest.java]: Fixed; thanks!

**Comment by risdenk** [comment]
Thanks @smolnar82
